### PR TITLE
@tommy/appjson schema sync

### PR DESCRIPTION
**Pull request by tommybru**
# Why

With an updated schema, we can now render a nicer App Configuration properties table automatically from exp.host

# How

- A new `schema-sync` file that can be run w/ `yarn run schema-sync 38` or `yarn run schema-sync unversioned`
- This will write a versioned copy of the schema pulled from the right exp.host version (or the UNVERSIONED schema manually set if the argv is `unversioned`)
- New `app-config.md` files that render custom component `AppConfigSchemaPropertiesTable`
- Renders a table with properties and subproperties, `type`, and `description` (which includes tips: Example strings, Bare Workflow/Expo Kit pointers, or a `regexHuman` description)

# Test Plan

`AppConfigSchemaPropertiesTable.test.js` currently tests:

- A snapshot test for the table styling (helps alert anyone to make sure all style changes still look good on a table as large as `AppConfigSchemaPropertiesTable`)
- A function-level test for `formatSchema` to make sure property levels are applied correctly, and
- A function-level test that `createDescription` applies each "tip" condition correctly (e.g., `regexHuman`, `exampleString` and `bareWorkflow` together in a property, `expoKit`)

### CODE REVIEW

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:30]:
```suggestion
    //The ` backticks are for markdown highlighting, and #### is to apply anchor linking only on top-level properties
```

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:53]:
```suggestion
function createDescription(property) {
```

`smartly` here does not really help me understand what this function does.

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:81]: This should be a function that returns the formatted schema, since `formattedSchema` is important to the data massaging functions, but not to this component.

Ideally this call would look something like this here in the `render` method:

```js
const formattedSchema = formatSchema(rawSchema);
```

then the formatSchema function would be something like:

```js
function formatSchema(rawSchema) {
  let formattedSchema = [];

  return rawSchema.map( ... )
}
```

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:18]:
```suggestion
function appendProperty(formattedSchema, property, _level) {
```

I'm a big fan of slightly longer function names to make them more readable. Up to you if you agree.

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:99]:
```suggestion
                      data-testid={property.name}
```

Why is `.match(/[a-zA-Z]+/g)` needed here? This value should be a string instead of an array

File: docs/components/plugins/AppConfigPropsTable.test.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.test.js:170]: 🔥 nice tests

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:73]:
```suggestion
export default class AppConfigSchemaPropertiesTable extends React.Component {
```

Possibly easier to understand name. `Props` to me means react props, but maybe im the only one.

File: docs/components/plugins/AppConfigPropsTable.js
Comment (jonsamp) [inline docs/components/plugins/AppConfigPropsTable.js:86]:
```suggestion
          <thead style={{ backgroundColor: expoColors.gray[100] }}>
```

Would be great to give this a bg color. This is imported from: `import { expoColors } from '~/common/constants';` --Might be worth putting this in a `css` variable, then using here as a `className`.

**Review by jonsamp — CHANGES_REQUESTED** [review]
Overall this is a massive improvement to our current docs. Nice work wrapping all this up.

I added a bunch of naming nit-picks. Those are optional to change.

The stuff I'd love to see fixed before this is merged is:
- When i navigate to "http://localhost:3000/versions/latest/sdk/app-config", I am redirected to "http://localhost:3000/versions/latest/sdk/app-config/?redirected" with the `?redirected` flag? I'm not sure where/how that's getting added?
- The test ids on the table should be strings instead of arrays.
- Id like the schema formatting logic to be wrapped in a function.

Let me know if you have questions. Im also happy to pair/talk about this more. Once you have changes, please re-request review from me!

File: docs/components/plugins/AppConfigPropsTable.js
Comment (tommybru) [reply docs/components/plugins/AppConfigPropsTable.js:99]: Currently this is because the property.name might equal something like: \`#### androidNavigationBar\` OR \`visible\`, so in the `.test` file I wanted to just be able to read these as just "androidNavigationBar" or "visible" for the reader's sake, but I can find another way to be clear

**Comment by tommybru** [comment]
Sg! And I think the reason for the `?redirected` flag getting added is because the url you provided doesn't end in a `/`

On localhost, I think we go through the extra step of showing you this redirect (no matter the link), although I think on the live docs.expo.io site, it's so quick that we usually hide the redirecting UI and `?redirected` flag

Screen Shot 2020-07-08 at 1 54 12 PM

File: docs/components/plugins/AppConfigPropsTable.test.js
Comment (tommybru) [reply docs/components/plugins/AppConfigPropsTable.test.js:170]: thank you !!

**Review by jonsamp — APPROVED** [review]

File: docs/components/plugins/AppConfigSchemaPropertiesTable.test.js
Comment (brentvatne) [inline docs/components/plugins/AppConfigSchemaPropertiesTable.test.js:114]: this seems a little bit fragile to test styles like this - we'd need to come and update this test any time we change them. an alternative way to do this without inlining the values is to use snapshots. another way to do this would be to test that there is some attribute set that is more generic that won't change, or that the component type is one that corresponds to a level 0 element

File: docs/components/plugins/AppConfigSchemaPropertiesTable.js
Comment (brentvatne) [inline docs/components/plugins/AppConfigSchemaPropertiesTable.js:24]:
```suggestion
  let formattedSchema = [];
```

File: docs/components/plugins/AppConfigSchemaPropertiesTable.js
Comment (brentvatne) [inline docs/components/plugins/AppConfigSchemaPropertiesTable.js:73]:
```suggestion
  let propertyDescription = propertyValue.description;
```

File: docs/components/plugins/AppConfigSchemaPropertiesTable.test.js
Comment (brentvatne) [inline docs/components/plugins/AppConfigSchemaPropertiesTable.test.js:141]: it may be easier to test `appendProperty` and `createDescription` directly - this would make it easier to work with this code in the future because they seem like the parts that we're most likely to change

File: docs/components/plugins/AppConfigSchemaPropertiesTable.test.js
Comment (tommybru) [reply docs/components/plugins/AppConfigSchemaPropertiesTable.test.js:114]: okay sounds good! do you know where I could get more information about snapshots?

And currently, the properties in `level_0_style` are all ones that I set within `AppConfigSchemaPropertiesTable.js` so I am not sure there is a more generic attribute set -> so snapshots might be the best solution right now

**Review by byCedric — APPROVED** [review]
Looks pretty solid! One thing I'm doubting is the sync-script to pull in the latest schema. In essence, we shouldn't need to manually update this. Once it's published on `www` (deployed), the manifest is in effect. But maybe that's a nice enhancement after this 😄

**Comment by tommybru** [comment]
Just updated the testing! Now it is:

- A snapshot test for the table styling (helps alert anyone to make sure all style changes still look good on a table as large as `AppConfigSchemaPropertiesTable`) 
- A function-level test for `formatSchema` to make sure property levels are applied correctly, and 
- A function-level test that `createDescription` applies each condition correctly

File: docs/components/plugins/AppConfigSchemaPropertiesTable.test.js
Comment (brentvatne) [reply docs/components/plugins/AppConfigSchemaPropertiesTable.test.js]: this is a good doc for that: https://jestjs.io/docs/en/snapshot-testing

you can also search `toMatchSnapshot()` in our codebase to find other examples of usage

**Review by brentvatne — APPROVED** [review]

**Pull request merged**
